src/raft.rs

```diff
@@ -1843,38 +1843,24 @@ impl<T: Storage> Raft<T> {
         self.prs().voter_ids().contains(&self.id)
     }
 
-    fn add_voter_or_learner(&mut self, id: u64, learner: bool) {
+    fn add_voter_or_learner(&mut self, id: u64, learner: bool) -> Result<()> {
         debug!(
             "Adding node (learner: {}) with ID {} to peers.",
             learner, id
         );
 
-        // Ignore redundant inserts.
-        // TODO: Remove these and have this function and related functions return errors.
-        if self.prs().get(id).is_some() {
-            // If progress is a learner, then it's already inserted as what it should be, return early to avoid error.
-            if self.prs().learner_ids().contains(&id) == learner {
-                info!("{} Ignoring redundant insert of ID {}.", self.tag, id);
-                return;
-            }
-            // If progress is a voter, and learner == true, then it's a demotion, return early to avoid an error.
-            if self.prs().voter_ids().contains(&id) && learner {
-                info!("{} Ignoring voter demotion of ID {}.", self.tag, id);
-                return;
-            }
-        };
-
-        let progress = Progress::new(self.raft_log.last_index() + 1, self.max_inflight);
         let result = if learner {
+            let progress = Progress::new(self.raft_log.last_index() + 1, self.max_inflight);
             self.mut_prs().insert_learner(id, progress)
         } else if self.prs().learner_ids().contains(&id) {
             self.mut_prs().promote_learner(id)
         } else {
+            let progress = Progress::new(self.raft_log.last_index() + 1, self.max_inflight);
             self.mut_prs().insert_voter(id, progress)
         };
 
-        if let Err(e) = result {
-            panic!("{}", e)
+        if let Err(ref e) = result {
+            error!("{}", e);
```

Comment:
Should return here. And seems a test case is missing.

Reply:
Ah, yes, you're very right.

Reply:
Any tests?

Reply:
I guess you mean to test that the `ProgressSet` is not mutated if an `Error` is returned?

I think it is better to test the `prs.insert_voter()` and `prs.insert_learner()` functions if that's the case.

Reply:
It's not just `ProgressSet`. For example, L1867 mutate the internal state of raft state machine.

Reply:
Where is the test case?

Reply:
I did not see this comment about L1867. Good catch. I'll add it.